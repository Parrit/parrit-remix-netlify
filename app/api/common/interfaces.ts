export interface LoginRequest {
  projectName?: string;
  password?: string;
}

export interface LoginResult {
  message: string;
}

export interface Project {
  id: string;
  name: string;
  pairingBoards: PairingBoard[];
  people: Person[];
}

export interface PairingBoard {
  id: string;
  name: string;
  exempt: boolean;
  people: Person[];
  roles: Role[];
}

export interface Person {
  id: string;
  name: string;
}

export interface Role {
  id: string;
  name: string;
}

export interface PairingArrangementDTO {
  id: number; // this should be changed to string
  pairingTime: string;
  pairingHistories: PairingHistory[];
}

export interface PairingHistory {
  pairingBoardName: string;
  people: Person[];
}
