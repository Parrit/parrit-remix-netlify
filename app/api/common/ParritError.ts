export interface IParritError<T> {
  server?: string;
  fields?: Record<keyof T, string | null>;
}

export class ParritError<T> extends Error {
  public readonly data: IParritError<T>;

  constructor(arg: IParritError<T>) {
    super(JSON.stringify(arg));
    this.data = arg;
  }

  toResponse(status = 400): Response {
    return new Response(JSON.stringify(this.data), { status });
  }

  static obscured() {
    return new ParritError({ server: "Server error" });
  }

  static fromString<T>(str?: string) {
    if (!str) {
      return undefined;
    }
    const obj = JSON.parse(str) as IParritError<T>;
    if (this.isParritError<T>) {
      return new ParritError(obj);
    }
    return undefined;
  }

  private static isParritError<T>(obj: object): obj is ParritError<T> {
    const cast = obj as ParritError<T>;
    return !!cast.data.server || !!cast.data.fields;
  }
}
