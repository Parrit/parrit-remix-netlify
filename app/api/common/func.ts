import {
  FLOATING_IDX,
  PairingBoard,
  Person,
  Project,
} from "./interfaces/parrit.interfaces";
import * as R from "ramda";

// Functional programing utilities for working with interfaces

export const movePerson = (
  project: Project,
  person: Person,
  position: PairingBoard
) =>
  R.pipe<[Project], Project, Project>(
    removePerson(person),
    addPerson(person, position)
  )(project);

export const addPerson = R.curry(
  (person: Person, position: PairingBoard, project: Project) => {
    const f1 = addPersonToPosition(person, position);
    const f2 =
      position.id === FLOATING_IDX ? addPersonToFloating(person) : R.identity;
    return R.pipe<[Project], Project, Project>(f1, f2)(project);
  }
);

const addPersonToPosition = R.curry(
  (person: Person, position: PairingBoard, project: Project) =>
    R.over<Project, PairingBoard[]>(
      R.lensProp("pairingBoards"),
      R.map<PairingBoard, PairingBoard>((x) => {
        if (x.id === position.id) {
          return R.over<PairingBoard, Person[]>(
            R.lensProp("people"),
            R.append({ ...person, pairing_board_id: position.id }),
            x
          );
        }
        return x;
      }),
      project
    )
);

const addPersonToFloating = R.curry((person: Person, project: Project) =>
  R.over<Project, Person[]>(
    R.lensPath(["floating", "people"]),
    R.append(person),
    project
  )
);

export const removePerson = R.curry(
  (person: Person, project: Project): Project =>
    R.pipe<[Project], Project, Project>(
      removePersonFromBoards(person),
      removePersonFromFloating(person)
    )(project)
);

const removePersonFromBoards = R.curry(
  (person: Person, project: Project): Project =>
    R.over<Project, PairingBoard[]>(
      R.lensProp("pairingBoards"),
      R.map<PairingBoard, PairingBoard>((x) => {
        if (x.people.includes(person)) {
          return R.set<PairingBoard, Person[]>(
            R.lensProp("people"),
            R.without(x.people, [person]),
            x
          );
        }
        return x;
      }),
      project
    )
);

const removePersonFromFloating = R.curry(
  (person: Person, project: Project): Project =>
    R.set<Project, Person[]>(
      R.lensPath(["floating", "people"]),
      R.without([person], project.floating.people),
      project
    )
);

const pbIndexById = R.curry((pbs: PairingBoard[], pb_id: string) =>
  R.findIndex<PairingBoard>((pb) => pb.id === pb_id)(pbs)
);

const renamePairingBoardFromList = R.curry(
  (pb_id: string, rename: string, pbs: PairingBoard[]) => {
    const foundIndex = pbIndexById(pbs, pb_id);
    if (foundIndex < 0) {
      return pbs;
    }
    return R.adjust<PairingBoard>(foundIndex, R.assoc("name", rename), pbs);
  }
);

export const renamePairingBoard = R.curry(
  (project: Project, pb_id: string, rename: string): Project =>
    R.over<Project, PairingBoard[]>(
      R.lensProp("pairingBoards"),
      renamePairingBoardFromList(pb_id, rename),
      project
    )
);

export const canAPairingBeMade = (project: Project): boolean => {
  const atLeast2Floaters = project.floating.people.length >= 2;
  const atLeast1Floater = project.floating.people.length >= 1;
  const atLeast1EmptyBoard = getEmptyPairingBoard(project) !== undefined;
  const atLeast1UnpairedSticker = unpairedStickingPeople(project).length >= 1;
  return (
    (atLeast1Floater && atLeast1UnpairedSticker) ||
    (atLeast2Floaters && atLeast1EmptyBoard)
  );
};

export const getEmptyPairingBoard = (
  project: Project
): PairingBoard | undefined => {
  return project.pairingBoards.find(
    (pb) => pb.people.length === 0 && !pb.exempt
  );
};

const unpairedStickingPeople = (project: Project): Person[] =>
  project.pairingBoards
    .filter((board) => !board.exempt)
    .filter((board) => board.people.length === 1)
    .flatMap((board) => board.people);

const findPairingBoardByPerson = (
  project: Project,
  person: Person
): PairingBoard | undefined =>
  project.pairingBoards.find((pb) => pb.people.find((p) => p.id === person.id));
