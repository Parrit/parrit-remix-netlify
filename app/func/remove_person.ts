import {
  curry,
  pipe,
  over,
  lensProp,
  map,
  set,
  without,
  lensPath,
} from "ramda";
import {
  Person,
  Project,
  PairingBoard,
} from "~/api/common/interfaces/parrit.interfaces";

export default curry(
  (person: Person, project: Project): Project =>
    pipe<[Project], Project, Project>(
      remove_person_from_boards(person),
      remove_person_from_floating(person)
    )(project)
);

const remove_person_from_boards = curry(
  (person: Person, project: Project): Project =>
    over<Project, PairingBoard[]>(
      lensProp("pairingBoards"),
      map<PairingBoard, PairingBoard>((board) => {
        if (board.people.map((p) => p.id).includes(person.id)) {
          return set<PairingBoard, Person[]>(
            lensProp("people"),
            without([person], board.people),
            board
          );
        }
        return board;
      }),
      project
    )
);

const remove_person_from_floating = curry(
  (person: Person, project: Project): Project =>
    set<Project, Person[]>(
      lensPath(["floating", "people"]),
      without([person], project.floating.people),
      project
    )
);
