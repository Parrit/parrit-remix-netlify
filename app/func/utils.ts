/**
 * These files are tested via the functionality of other tests.
 * That's what makes them utils and not their own functions
 */

import {
  Project,
  Person,
  PairingBoard,
} from "~/api/common/interfaces/parrit.interfaces";

export const can_a_pairing_be_made = (project: Project): boolean => {
  const atLeast2Floaters = project.floating.people.length >= 2;
  const atLeast1Floater = project.floating.people.length >= 1;
  const atLeast1EmptyBoard = get_empty_pairing_board(project) !== undefined;
  const atLeast1UnpairedSticker = unpaired_sticking_people(project).length >= 1;
  return (
    (atLeast1Floater && atLeast1UnpairedSticker) ||
    (atLeast2Floaters && atLeast1EmptyBoard)
  );
};

export const unpaired_sticking_people = (project: Project): Person[] =>
  project.pairingBoards
    .filter((board) => !board.exempt)
    .filter((board) => board.people.length === 1)
    .flatMap((board) => board.people);

export const get_empty_pairing_board = (
  project: Project
): PairingBoard | undefined => {
  return project.pairingBoards.find(
    (pb) => pb.people.length === 0 && !pb.exempt
  );
};

export const all_people_in_project = (project: Project): Person[] => {
  const people: Person[] = [...project.floating.people];
  project.pairingBoards.forEach((pb) =>
    pb.people.forEach((person) => people.push(person))
  );
  return people;
};

export const find_pairing_board_by_person = (
  project: Project,
  person: Person
): PairingBoard | undefined =>
  project.pairingBoards.find((pb) => pb.people.find((p) => p.id === person.id));
