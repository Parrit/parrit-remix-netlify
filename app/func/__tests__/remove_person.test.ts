import { expect } from "@jest/globals";
import { RED_PERSON, PROJECT_START, FLOATING_PERSON } from "./common";
import remove_person from "../remove_person";
import { can_a_pairing_be_made } from "../utils";
import {
  FLOATING_IDX,
  Project,
} from "~/api/common/interfaces/parrit.interfaces";

describe("remove_person", () => {
  it("removing from pb", () => {
    const result = remove_person({ ...RED_PERSON }, PROJECT_START);
    expect(result.pairingBoards[0].people.length).toBe(0);
    expect(result).toMatchSnapshot();
    expect(can_a_pairing_be_made(result)).toBe(false); // only one person in floating, no unpaired assigned
  });

  it("removing from floating", () => {
    const result = remove_person({ ...FLOATING_PERSON }, PROJECT_START);
    expect(result).toMatchSnapshot();
    expect(can_a_pairing_be_made(result)).toBe(false); // no floating pairs
  });

  it("problematic input", () => {
    const before: Project = {
      floating: {
        name: "",
        id: FLOATING_IDX,
        exempt: false,
        roles: [],
        people: [
          {
            id: "20233",
            name: "David",
            type: "Person",
            pairing_board_id: "-1",
          },
          {
            id: "19093",
            name: "Billy",
            type: "Person",
            pairing_board_id: "-1",
          },
          { id: "19092", name: "John", type: "Person", pairing_board_id: "-1" },
          {
            id: "rec_ct6kg2ve4lpv8v5i3200",
            name: "Mario",
            type: "Person",
            pairing_board_id: "-1",
          },
          {
            id: "rec_ct8k0bfhvh3f7vsh7050",
            name: "jimbo",
            type: "Person",
            pairing_board_id: "-1",
          },
          {
            id: "rec_ct8k0bvhvh3f7vsh705g",
            name: "jimbo",
            type: "Person",
            pairing_board_id: "-1",
          },
          {
            id: "rec_ct8k0bu44ggk3j1pqng0",
            name: "jimbo",
            type: "Person",
            pairing_board_id: "-1",
          },
        ],
      },
      id: "",
      name: "",
      pairingBoards: [],
    };

    const after = remove_person(
      {
        id: "rec_ct8k0bu44ggk3j1pqng0",
        name: "jimbo",
        type: "Person",
        pairing_board_id: "-1",
      },
      before
    );

    expect(after.floating.people.map((p) => p.id)).not.toContain(
      "rec_ct8k0bu44ggk3j1pqng0"
    );
    expect(after).toMatchSnapshot();
  });
});
