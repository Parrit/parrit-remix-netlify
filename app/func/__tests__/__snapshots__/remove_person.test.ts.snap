// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`remove_person problematic input 1`] = `
{
  "floating": {
    "exempt": false,
    "id": "-1",
    "name": "",
  },
  "id": "",
  "name": "",
  "pairingBoards": [],
  "people": [
    {
      "id": "20233",
      "name": "David",
      "pairing_board_id": "-1",
      "type": "Person",
    },
    {
      "id": "19093",
      "name": "Billy",
      "pairing_board_id": "-1",
      "type": "Person",
    },
    {
      "id": "19092",
      "name": "John",
      "pairing_board_id": "-1",
      "type": "Person",
    },
    {
      "id": "rec_ct6kg2ve4lpv8v5i3200",
      "name": "Mario",
      "pairing_board_id": "-1",
      "type": "Person",
    },
    {
      "id": "rec_ct8k0bfhvh3f7vsh7050",
      "name": "jimbo",
      "pairing_board_id": "-1",
      "type": "Person",
    },
    {
      "id": "rec_ct8k0bvhvh3f7vsh705g",
      "name": "jimbo",
      "pairing_board_id": "-1",
      "type": "Person",
    },
  ],
  "roles": [],
}
`;

exports[`remove_person removing from floating 1`] = `
{
  "floating": {
    "exempt": false,
    "id": "-1",
    "name": "floating_board_name",
  },
  "id": "project_id",
  "name": "project_name",
  "pairingBoards": [
    {
      "exempt": false,
      "id": "red_id",
      "name": "red_start",
    },
    {
      "exempt": false,
      "id": "blue_id",
      "name": "blue_start",
    },
    {
      "exempt": true,
      "id": "exempt",
      "name": "out_of_office",
    },
  ],
  "people": [
    {
      "id": "red_parrit_id",
      "name": "red_parrit",
      "pairing_board_id": "red_id",
      "type": "Person",
    },
    {
      "id": "99",
      "name": "Ms. Frizzle",
      "pairing_board_id": "brand new individual",
      "type": "Person",
    },
    {
      "id": "absent",
      "name": "Permanent Vacation",
      "pairing_board_id": "exempt",
      "type": "Person",
    },
  ],
  "roles": [
    {
      "id": "blue_role",
      "name": "Blue Role",
      "pairing_board_id": "blue_id",
      "type": "Role",
    },
    {
      "id": "ooo",
      "name": "OOOROLE",
      "pairing_board_id": "exempt",
      "type": "Role",
    },
  ],
}
`;

exports[`remove_person removing from pb 1`] = `
{
  "floating": {
    "exempt": false,
    "id": "-1",
    "name": "floating_board_name",
  },
  "id": "project_id",
  "name": "project_name",
  "pairingBoards": [
    {
      "exempt": false,
      "id": "red_id",
      "name": "red_start",
    },
    {
      "exempt": false,
      "id": "blue_id",
      "name": "blue_start",
    },
    {
      "exempt": true,
      "id": "exempt",
      "name": "out_of_office",
    },
  ],
  "people": [
    {
      "id": "99",
      "name": "Ms. Frizzle",
      "pairing_board_id": "brand new individual",
      "type": "Person",
    },
    {
      "id": "absent",
      "name": "Permanent Vacation",
      "pairing_board_id": "exempt",
      "type": "Person",
    },
    {
      "id": "floater_id",
      "name": "floater_name_start",
      "pairing_board_id": "-1",
      "type": "Person",
    },
  ],
  "roles": [
    {
      "id": "blue_role",
      "name": "Blue Role",
      "pairing_board_id": "blue_id",
      "type": "Role",
    },
    {
      "id": "ooo",
      "name": "OOOROLE",
      "pairing_board_id": "exempt",
      "type": "Role",
    },
  ],
}
`;
